janitor/functions/conditional_join.py

```diff
@@ -1151,9 +661,493 @@ def _not_equal_indices(
+        )
+        if outcome is None:
+            return None
+
+        return _generate_indices(*outcome, conditions)
+
+
+def _range_indices(
+    df: pd.DataFrame,
+    right: pd.DataFrame,
+    first: tuple,
+    second: tuple,
+    rest: tuple = None,
+):
+    """
+    Retrieve index positions for range/interval joins.
+
+    Idea inspired by article:
+    https://www.vertica.com/blog/what-is-a-range-join-and-why-is-it-so-fastba-p223413/
+
+    Returns a tuple of (left_index, right_index)
+    """
+
+    # summary of code for range join:
+    # get the positions where start_left is >/>= start_right
+    # then within the positions,
+    # get the positions where end_left is </<= end_right
+    # this should reduce the search space
+    left_on, right_on, op = first
+
+    strict = False
+    if op == _JoinOperator.GREATER_THAN.value:
+        strict = True
+
+    outcome = _greater_than_indices(
+        df[left_on], right[right_on], strict, multiple_conditions=True
+    )
+
+    if outcome is None:
+        return None
+
+    left_index, right_index, search_indices = outcome
+    left_on, right_on, op = second
+    right_c = right.loc[right_index, right_on]
+    left_c = df.loc[left_index, left_on]
+    left_c = extract_array(left_c, extract_numpy=True)
+    op = operator_map[op]
+    pos = np.copy(search_indices)
+    counter = np.arange(left_index.size)
+    ext_arr = is_extension_array_dtype(left_c)
+
+    dupes = right_c.duplicated(keep="first")
+    right_c = extract_array(right_c, extract_numpy=True)
+    # use position, not label
+    uniqs_index = np.arange(right_c.size)
+    if dupes.any():
+        uniqs_index = uniqs_index[~dupes]
+        right_c = right_c[~dupes]
+
+    if ext_arr:
+        for ind in range(uniqs_index.size):
+            keep_rows = op(left_c, right_c[ind])
+            keep_rows = keep_rows.to_numpy(dtype=bool, na_value=False)
+            # get the index positions where left_c is </<= right_c
+            # that minimum position combined with the equivalent position
+            # from search_indices becomes our search space
+            # for the equivalent left_c index
+            if keep_rows.any():
+                pos[counter[keep_rows]] = uniqs_index[ind]
+                counter = counter[~keep_rows]
+                left_c = left_c[~keep_rows]
+            if not counter.size > 0:
+                break
+    else:
+        for ind in range(uniqs_index.size):
+            keep_rows = op(left_c, right_c[ind])
+            # get the index positions where left_c is </<= right_c
+            # that minimum position combined with the equivalent position
+            # from search_indices becomes our search space
+            # for the equivalent left_c index
+            if keep_rows.any():
+                pos[counter[keep_rows]] = uniqs_index[ind]
+                counter = counter[~keep_rows]
+                left_c = left_c[~keep_rows]
+            if not counter.size > 0:
+                break
```

Comment:
if the column is an extension array (Int64, Float64, categoricals), the numpy array wont be extracted (numpy does not recognize pd.NA for instance). `extract_array` will only return numpy arrays if the data is not an extension array.

Reply:
i'd probably be more strict and just avoid extension arrays, similar to what pd.merge_asof does; i'm just not sure it is the right approach in terms of restrictions. i am open to suggestions

Reply:
I see. Thanks, didn't know about that behaviour of `extract_array`.

And when you compare two extension arrays (`op(left_c, right_c)`), is there a situation where you get an extension array back? -- The reason I'm asking is because that `to_numpy()` call is the only thing different between `ext_arr=True/False`. --
From my very limited testing, it seems if I compare 2 Categorical extensiondtype arrays, I get a numpy boolean array back.

```python
from pandas.core.construction import extract_array
from pandas.core.dtypes.common import is_extension_array_dtype
import operator

s1 = pd.Categorical(pd.Series([1,2,3.,pd.NA]), ordered=True)
s2 = pd.Categorical(pd.Series([2,3,1.,pd.NA]), ordered=True)
assert is_extension_array_dtype(s1)
assert is_extension_array_dtype(s2)

s1 = extract_array(s1, extract_numpy=True)
s2 = extract_array(s2, extract_numpy=True)

res = operator.le(s1, s2)
display(res)  # array([ True,  True, False, False])
is_extension_array_dtype(res)  # False
```

Reply:
How about just:
1. not bothering to extract_array / numpy at the beginning of the function (and removing the current `if ext_arr` check)
2. Then do the if-else check inside:
    ```python
    for ind in range(uniqs_index.size):
        keep_rows = op(left_c, right_c[ind])
        if keep_rows.any():
            # [[ move the is extension array check here ]]
            if is_extension_array_dtype(keep_rows):
                keep_rows = keep_rows.to_numpy(dtype=bool, na_value=False)
            pos[counter[keep_rows]] = uniqs_index[ind]
            counter = counter[~keep_rows]
            left_c = left_c[~keep_rows]
        if not counter.size > 0:
            break
    ```

Would this break anything?

Reply:
so in pandas 1.3.5 (which is the version we r using in the config - i am working in the docker env) , a pandas BooleanArray is returned, not an ndarray. And thanks for the observations, apparently categories are coerced to ndarrays, when they are compared :

```py
In [20]: import operator
    ...: import pandas as pd
    ...: from pandas.core.construction import extract_array
    ...: from pandas.core.dtypes.common import is_extension_array_dtype
    ...: 
    ...: 
    ...: s1 = pd.Series([1, 2, 3.0, pd.NA], dtype="Int64")
    ...: s2 = pd.Series([2, 3, 1.0, pd.NA], dtype="Int64")
    ...: assert is_extension_array_dtype(s1)
    ...: assert is_extension_array_dtype(s2)
    ...: 
    ...: s1 = extract_array(s1, extract_numpy=True)
    ...: s2 = extract_array(s2, extract_numpy=True)
    ...: 
    ...: res = operator.le(s1, 2)
    ...: display(res)  # array([ True,  True, False, False])
    ...: is_extension_array_dtype(res)  # False
<BooleanArray>
[True, True, False, <NA>]
Length: 4, dtype: boolean
Out[20]: True
```

I think I moved it out because I didnt want to pay the price of checking for the extension array each time (imagine so many iterations and for each one a check for extension array dtype), compared to doing the check only once (hence the duplicated code). still open to a more performant option though.


do you mind explaining the logic for the guard clause (for the existing code, if keep_rows.any() is False, nothing is executed right?)

pls, i appreciate this feedback and the hole poking (i feel bluntness is better for growth), so pls bring them on. this is a sure fire way for me to improve. so thank you!

Reply:
1. Ah yes, indeed. I kinda figured the BooleanArray (extension array) would pop up somewhere after I wrote the initial comment. Should've played around with more examples 😅.
2. Okay, I see your point about the if-check happening O(n) times -- I had just felt that an additional if-check on a precomputed boolean variable shouldn't have that much of an effect on the runtime since the loop contains a more expensive computation. I got these results while I was experimenting (though it's probably not the right way to test, hah. 🤷🏼‍♂️) 
```python
import numpy as np
rng = np.random.default_rng(seed=42)
df = pd.DataFrame(rng.integers(0, 100, size=(50_000, 2)), columns=['a', 'b'])
s1, s2 = df['a'], df['b']
ext_arr = True

%%timeit -n3 -r5
for ind in range(len(s2)):
    keep_rows = s1> s2[ind]  # "expensive" computation
    if ext_arr: pass  # comment this out (?)

# with if-check:      6.89 s ± 380 ms per loop (mean ± std. dev. of 5 runs, 3 loops each)
# without if-check: 6.54 s ± 228 ms per loop (mean ± std. dev. of 5 runs, 3 loops each)
```
Regardless, I think it's largely just a matter of trade-off. Could we get one more opinion on this (@ericmjl ? 😄 ) and then I'll let this matter rest regardless, haha. Probably the more performant option will come from dropping down to numba so.. I'll still lean towards having cleaner code and DRY (don't-repeat-yourself) instead of having code duplication.

3. With respect to guard clauses, e.g. pattern described in general [here](https://medium.com/lemon-code/guard-clauses-3bc0cd96a2d3). You're right in saying the code in the if-block doesn't get executed, so performance- and functionality-wise, they are equivalent. It's more for ensuring code readability, as the "main" logic (when `keep_rows.any() == True`) is left un-nested.
```python
for ind in range(uniqs_index.size):
    if not counter.size > 0:
        # this is actually checking the counter of the previous iteration (or, of the initial counter
        # before the loop started). It is 99% equivalent to having the check at the end of the
        # loop (the way it is now). If this check is placed first, we (a) exit the loop if the original
        # counter is alr empty; but (b) lose a "break" statement if the counter happens to be
        # empty at the end of the loop. But this is inconsequential since the loop has already
        # ended.
        break
    if not keep_rows.any():
        # this is the main guard clause logic. It is just the inverse of the existing logic, so the
        # code after this block is NOT executed if keep_rows.any() is False. Functionally, it
        # makes no difference whether you do it this way or leave it the way it is now, but...
        continue
    # the benefit is this code block (which is where the loop will end up most of the time,
    # and contains the "heaviest" logic) is unnested. This gives you more "room" to work
    # with before reaching the PEP8 char limit; and I don't know about you guys, but I tend
    # to skip over all the if-checks when casting my eye over code.
    pos[counter[keep_rows]] = uniqs_index[ind]
    counter = counter[~keep_rows]
    left_c = left_c[~keep_rows]
```

Reply:
I guess I have to adjudicate? :joy_cat:

Regarding guard clauses: The pattern is a good one, definitely for readability and maintainability. Thanks, @thatlittleboy for finding the definition and explaining it. I now have another word in my vocabulary to name the good programming pattern. @samukweku, would you be open to modifying your implementation? It'd definitely be an upgrade for readability.

Regarding the point I was actually asked to adjudicate :joy_cat:, I actually am a bit lost. (ahhhh!) Is it the case that the guard clause pattern induces the extra check? Without the details in my mental RAM, I'd default to the following principles to decide:

1. Reduce duplication,
2. Increase readability,
3. Document extensively and locally,
4. Flat is better than nested,
5. If there's little practical difference in runtime, then I guess we're okay :smile:.

Reply:
😂 The simple tl;dr is: 1) We have a codeblock that has largely duplicated logic (the highlighted code block at the very beginning of this thread). 2) I was making the proposal to shift the `if ext_arr` logic *inside* the for-loop, so we can avoid the duplication of code (a comparison between the if- and else- block shows you the only difference is the conversion of the extension array into a numpy array).

So either we: 1) leave it as it is -- duplicated logic, but do if check only once outside the loop, or 2) shift the if check inside resulting in O(n) number of if checks (possibly slower, but I doubt so in any meaningful fashion), but we achieve DRY-er code.